k8sapiserver/k8sapiserver.go

```diff
@@ -81,199 +74,167 @@ func StartAPIServer(kubeAPIServerURL, etcdURL, frontendURL string) (*restclient.
+		}
+	}()
 
-	return cfg
+	return closeFn, nil
 }
 
-type fakeLocalhost443Listener struct{}
+func createK8SAPIServerOpts(etcdURL, frontendURL string) (*apiserverappopts.ServerRunOptions, func(), error) {
+	serverOpts := apiserverappopts.NewServerRunOptions()
 
-func (fakeLocalhost443Listener) Accept() (net.Conn, error) {
-	return nil, nil
-}
+	// set up etcd
+	serverOpts.Etcd.StorageConfig.Transport.ServerList = []string{etcdURL}
 
-func (fakeLocalhost443Listener) Close() error {
-	return nil
-}
+	// set up CORS
+	serverOpts.GenericServerRunOptions.CorsAllowedOriginList = []string{frontendURL}
 
-func (fakeLocalhost443Listener) Addr() net.Addr {
-	return &net.TCPAddr{
-		IP:   net.IPv4(127, 0, 0, 1),
-		Port: 443,
+	// set up RBAC authorization and annoymous auth.
+	serverOpts.Authorization.Modes = []string{authzmodes.ModeRBAC}
+	serverOpts.Authentication.Anonymous.Allow = true
+	err := serverOpts.APIEnablement.RuntimeConfig.Set("api/all=true")
+	if err != nil {
+		return nil, nil, xerrors.Errorf("k8s api server set runtime config: %w", err)
 	}
-}
 
-// startAPIServer starts a kubernetes API server and an httpserver to handle api requests.
-//nolint:funlen
-func startAPIServer(controlPlaneConfig *controlplane.Config, s *httptest.Server, apiServerReceiver *APIServerHolder) (*controlplane.Instance, *httptest.Server, func(), error) {
-	var m *controlplane.Instance
-
-	stopCh := make(chan struct{})
-	closeFn := func() {
-		if m != nil {
-			if err := m.GenericAPIServer.RunPreShutdownHooks(); err != nil {
-				klog.Errorf("failed to run pre-shutdown hooks for api server: %v", err)
-			}
-		}
-		close(stopCh)
-		s.Close()
+	// setup fake key for secure serving
+	saSigningKeyFile, err := ioutil.TempFile("/tmp", "insecure_test_key")
+	if err != nil {
+		return nil, nil, xerrors.Errorf("create temp file failed: %v", err)
+	}
+	cleanupFunc := func() {
+		os.RemoveAll(saSigningKeyFile.Name())
 	}
 
-	clientset, err := clientset.NewForConfig(controlPlaneConfig.GenericConfig.LoopbackClientConfig)
-	if err != nil {
-		return nil, nil, nil, xerrors.Errorf("create clientset: %w", err)
+	if err = ioutil.WriteFile(saSigningKeyFile.Name(), []byte(ecdsaPrivateKey), 0o600); err != nil {
+		return nil, nil, xerrors.Errorf("write file %s failed: %v", saSigningKeyFile.Name(), err)
 	}
+	serverOpts.ServiceAccountSigningKeyFile = saSigningKeyFile.Name()
+	serverOpts.Authentication.ServiceAccounts.Issuers = []string{"https://foo.bar.example.com"}
+	serverOpts.Authentication.ServiceAccounts.KeyFiles = []string{saSigningKeyFile.Name()}
 
-	controlPlaneConfig.ExtraConfig.VersionedInformers = informers.NewSharedInformerFactory(clientset, controlPlaneConfig.GenericConfig.LoopbackClientConfig.Timeout)
+	// disable admission plugins to avoid node taints, service account, cert approval etc.
+	serverOpts.Admission.GenericAdmission.DisablePlugins = apiserveropts.AllOrderedPlugins
+	serverOpts.Admission.GenericAdmission.EnablePlugins = []string{}
 
```

Comment:
Can we avoid file permission issue with `serverOpts.SecureServing.ServerCert.CertDirectory` config like this?:
```suggestion

	dir, err := os.MkdirTemp("", "apiserver-certs")
	if err != nil {
		return nil, fmt.Errorf("create temp dir: %w", err)
	}
	serverOpts.SecureServing.ServerCert.CertDirectory = dir

```

Reply:
Got it, thanks for the suggestion !